internal/credentials/credentialls.go

```diff
@@ -0,0 +1,176 @@
+	//		Type: corev1.SecretTypeOpaque,
+	data := map[string][]byte{
+		"kubeconfig": kubeconfigBytes,
+	}
+	return r.createOrUpdateClusterCredentialSecret(ctx, cd, KubeconfigSecretName(cd.Name), data, kubeconfig)
+}
+
+func (r *Credentials) EnsureAdminPasswordSecret(ctx context.Context, cd *hivev1.ClusterDeployment) (string, error) {
+	secretRef := types.NamespacedName{Namespace: cd.Namespace, Name: KubeadminPasswordSecretName(cd.Name)}
+	secret := &corev1.Secret{}
+	err := r.Get(ctx, secretRef, secret)
+	if err == nil {
+		password, exists := secret.Data["password"]
+		if !exists {
+			r.Log.Warn("failed to find password in secret, generating new one")
+		} else {
+			passwordHash, err := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
+			if err != nil {
+				return "", fmt.Errorf("failed to generate password hash: %w", err)
+			}
+			return string(passwordHash), nil
+		}
+	} else if !errors.IsNotFound(err) {
+		return "", fmt.Errorf("failed to get kubeadmin password secret: %w", err)
+	}
+	r.Log.Infof("Generating admin password")
+	password, err := generateKubeadminPassword()
+	if err != nil {
+		return "", fmt.Errorf("failed to generate password: %w", err)
+	}
+	data := map[string][]byte{
+		"username": []byte(DefaultUser),
+		"password": []byte(password),
+	}
+
+	if err := r.createOrUpdateClusterCredentialSecret(ctx, cd, KubeadminPasswordSecretName(cd.Name), data, kubeadmincreds); err != nil {
+		return "", fmt.Errorf("failed to create kubeadmin password secret: %w", err)
+	}
+	passwordHash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate password hash: %w", err)
+	}
+	return string(passwordHash), nil
+}
+
+func (r *Credentials) createOrUpdateClusterCredentialSecret(ctx context.Context, cd *hivev1.ClusterDeployment, name string, data map[string][]byte, secretType string) error {
+	secret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cd.Namespace,
+		},
+		Data: data,
+	}
+
+	secret.Labels = AddLabel(secret.Labels, "hive.openshift.io/cluster-deployment-name", cd.Name)
+	secret.Labels = AddLabel(secret.Labels, "hive.openshift.io/secret-type", secretType)
+
+	deploymentGVK, err := apiutil.GVKForObject(cd, r.Scheme)
+	if err != nil {
+		r.Log.WithError(err).Errorf("error getting GVK for clusterdeployment")
+		return err
+	}
+	secret.OwnerReferences = []metav1.OwnerReference{{
+		APIVersion:         deploymentGVK.GroupVersion().String(),
+		Kind:               deploymentGVK.Kind,
+		Name:               cd.Name,
+		UID:                cd.UID,
+		BlockOwnerDeletion: pointer.Bool(true),
+	}}
+	mutateFn := func() error {
+		// Update the Secret object with the desired data
+		secret.Data = data
+		return nil
+	}
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, secret, mutateFn)
```

Comment:
I feel like the controller should be the one interacting with the API objects.

This package should probably just be creating the credential values similar to the separation we have in the certs package.

Reply:
I wanted to move everything related to the cluster credentials from the controller to this package (including the API objects).
I even considered moving the method that generates the password to another package (utils.go?)
Let's discuss it on Monday.